src/lib/Inference.hs

```diff
@@ -556,6 +518,33 @@ bindPat' (WithSrc pos pat) val = addSrcContext pos $ case pat of
     return $ env1 <> env2
   UPatVariant _ _ _ -> throw TypeErr "Variant not allowed in can't-fail pattern"
   UPatVariantLift _ _ -> throw TypeErr "Variant not allowed in can't-fail pattern"
+  UPatTable ps -> do
+    concreteIdxTy <- case getType val of
```

Comment:
Hmmm this makes me very uneasy, because it can make the type inference depend on the program order. This is because there will be situations where the type of `val` has been constrained by some earlier expressions, and there will be situations where it is only going to get constrained later. Rearranging independent expressions shouldn't change the conclusion of the type checker, but this looks like it could cause that behavior.

Note that when doing inference for table literals we don't do the check based on the zonked type suggestion, but only take the suggestion into account when it has been guaranteed to contain no free variables by the inference mode we're in (Concrete).

Reply:
Yeah, I guess this could lead to confusing behavior. Should I just restrict to unpacking `Fin n` tables for now, or is there a better fix here? I don't think we have an inference mode when processing patterns (although I guess we could add one...?)

Reply:
Restricting to only `Fin n` for now, since that seems the most immediately useful.